compiler/src/org.graalvm.compiler.core.aarch64/src/org/graalvm/compiler/core/aarch64/AArch64ArithmeticLIRGenerator.java

```diff
@@ -506,9 +507,23 @@ public Variable emitLoad(LIRKind kind, Value address, LIRFrameState state) {
     }
 
     @Override
-    public void emitStore(ValueKind<?> lirKind, Value address, Value inputVal, LIRFrameState state) {
-        AArch64AddressValue storeAddress = getLIRGen().asAddressValue(address);
+    public Variable emitVolatileLoad(LIRKind kind, Value address, LIRFrameState state) {
+        AllocatableValue loadAddress = asAllocatable(address);
```

Comment:
Does it need to ensure the address as a base register one for volatile access? See the "ensureBaseRegisterAddress()" in https://github.com/oracle/graal/pull/313

Reply:
Sorry I didn't get back to you earlier. This PR is on hold until the related bug https://github.com/oracle/graal/pull/1844 is fixed. Also, as Andrew mentioned, unsafe volatile accesses handling is missing from https://github.com/oracle/graal/pull/1642. That needs to be fixed before this PR is integrated. I have a patch that I haven't submitted for this. But again I need https://github.com/oracle/graal/pull/1844 fixed first.

Reply:
Thank you so much for the updating! I have a patch to handle the unsafe atomics. So with your patch, do you mean we use the "LateMembar" option on AArch64 to decide whether to use LDAR/STLR for java volatile access? If so, we should also use this option to handle the unsafe atomic. But in my patch I use the jvm option "-XX:+UseBarriersForVolatile" as the switch. Because whatever the option value is, the unsafe atomic always generate a single op. And it's strange to use "LateMembar" for it. So do you have any better idea or option to handle this? Actually we should use the same option to control Java volatile access and unsafe atomics.

Reply:
Does the "LateMembar" option also work for SubstrateVM?

Reply:
> Does the "LateMembar" option also work for SubstrateVM?

I'm not familiar enough with SVM to answer that question

Reply:
To clarify, the thing that I have a patch for is Unsafe.get*Volatile/Unsafe.put*Volatile. You're right that hotspot's option UseBarriersForVolatile should be factored int. I'm unsure how to handle this but I'll make sure I'll take it into account when I rework that patch.

Reply:
@christianwimmer can you please address the SVM question above.

Reply:
@rwestrel  I have a PR (https://github.com/oracle/graal/pull/2181) for the unsafe atomic issues. I only used the `"-XX:+UseBarriersForVolatile"` to decide the codes for atomics. And I think maybe I should also consider the graal option `“LateMambars”`, so you think so? And I also adds a same option for substratevm.  And anyway the options to decide codes for Java volatile access and unsafe atomics should keep the same. So if you have any better idea to handle this, please let me know. Thanks!

Reply:
> Does it need to ensure the address as a base register one for volatile access? See the "ensureBaseRegisterAddress()" in #313

I'm confused by this. Isn't:
@Use protected AllocatableValue address;
sufficient?
@adinn: what do you think?

Reply:
> @rwestrel I have a PR (#2181) for the unsafe atomic issues. It only used the `"-XX:+UseBarriersForVolatile"` to decide the codes for atomics. And I think maybe I should also consider the graal option `“LateMambars”`, do you think so? And I also adds the same option for substratevm. And anyway the options to decide codes for Java volatile access and unsafe atomics should keep the same. So if you have any better idea to handle this, please let me know. Thanks!

LateMembars got removed recently. The updated I push yesterday relies on UseBarriersForVolatile.

Reply:
Yes, cool! Actually there are some controversy for the vm option "-XX:+UseBarriersForVolatile" (See thread https://mail.openjdk.java.net/pipermail/aarch64-port-dev/2020-March/008679.html and https://mail.openjdk.java.net/pipermail/aarch64-port-dev/2020-April/008706.html). Although the final resolution is in discussion now,  this flag `UseBarriersForVolatile` might be removed in Graal in future.

Reply:
> > Does it need to ensure the address as a base register one for volatile access? See the "ensureBaseRegisterAddress()" in #313
> 
> I'm confused by this. Isn't:
> @use protected AllocatableValue address;
> sufficient?
> @adinn: what do you think?

The method ensureBaseRegisterAddress() that XiaohonGong is referring to from my earlier patch deals with the restricted addressing modes available when using ldar/stlr. If the address is not actually a plain BASE_REGISTER_ONLY address then an lea is planted (it is actually coded as a Move insn) and the target of that lea is passed to the ladr/stlr. I think he may be pointing out the need to do this?

Reply:
Right. But isn't that already guaranteed by
```
@Use protected AllocatableValue address;
```
where
```
    public static @interface Use {

        OperandFlag[] value() default OperandFlag.REG;
    }
```
? Doesn't that tell the register allocator that the address input needs to be in a register?

Reply:
Hmm, I think you must be mistaking me for a compiler expert :-)

Reply:
Basically there are several address modes in AArch64. The `address` here is actually a `AArch64AddressValue`, which might has other address modes except `BASE_REGISTER_ONLY`. You can have a look at `AArch64AddressValue` and  `AArch64Address.AddressingMode`.

Reply:
```
    public static final class VolatileLoadOp extends AArch64LIRInstruction {
        public static final LIRInstructionClass<VolatileLoadOp> TYPE = LIRInstructionClass.create(VolatileLoadOp.class);
        protected final AArch64Kind kind;
        @State protected LIRFrameState state;
        @Def protected AllocatableValue result;
        @Use protected AllocatableValue address;
```
It's not an AArch64AddressValue, it's AllocatableValue so I think it's guaranteed to be assigned a register.

Reply:
Yes, you are right here, Roland (it's such a while since I looked at this I I had to reread the code to be sure). AArch64AddressValue is a Value but is not an instance of AllocatableValue. So, asAllocatable will detect that and plant a move (it calls emitMove) returning a Register as the AllocatableValue passed to ldar/stlr.

Reply:
Thanks for the confirmation that's your understanding too.

Reply:
Hi @rwestrel , the VM option "UseBarriersForVolatile" is finally to be determined to obsolete in jdk (See: https://bugs.openjdk.java.net/browse/JDK-8243339). I think we should remove it for Graal either.  I will also update my Unsafe atomic PR by removing it in future. Thanks!

Reply:
@XiaohongGong do you think I should update this patch to remove UseBarriersForVolatile now or do we want to wait for JDK-8243339 to be pushed (and backported?) first?

Reply:
I think we can remove it now, since it is not actually used ever before in Graal. Thanks!

Reply:
I just pushed an updated patch with no reference to UseBarriersForVolatile

Reply:
OK！Thanks so much, it looks ok to me.